itou/utils/triggers/__init__.py

```diff
@@ -0,0 +1,113 @@
+import contextlib
+import functools
+import json
+import operator
+import threading
+from typing import Any
+
+from django.db import connection, models
+from pgtrigger import Condition, core
+
+
+_context = threading.local()
+_context_is_up_to_date = threading.Event()
+_context_query_in_progress = threading.Event()
+_context_connection_wrapper_installed = threading.Event()
+
+
+def _set_context_connection_wrapper(execute, *args, **kwargs):
+    if not _context_is_up_to_date.is_set() and not _context_query_in_progress.is_set():
+        with connection.cursor() as cursor:
+            _context_query_in_progress.set()
+            try:
+                cursor.execute(
+                    "SELECT set_config('itou.context', %s, true)",
+                    [json.dumps(_context.data)],
+                )
+            finally:
+                _context_query_in_progress.clear()
+        _context_is_up_to_date.set()
+
+    return execute(*args, **kwargs)
+
+
+@contextlib.contextmanager
+def context(**kwargs):
+    previous_data, _context.data = (
+        getattr(_context, "data", None),
+        kwargs,
+    )  # FIXME: Should we merge instead of replace?
+    _context_is_up_to_date.clear()
+
+    with contextlib.ExitStack() as stack:
+        if not _context_connection_wrapper_installed.is_set():
+            stack.enter_context(connection.execute_wrapper(_set_context_connection_wrapper))
+            _context_connection_wrapper_installed.set()
+            stack.callback(_context_connection_wrapper_installed.clear)
```

Comment:
Alternative évitant un `Event`:
```suggestion
        if _set_context_connection_wrapper not in connection.execute_wrappers:
            stack.enter_context(connection.execute_wrapper(_set_context_connection_wrapper))
```